core/src/main/java/com/linecorp/armeria/server/Routers.java

```diff
@@ -216,18 +218,45 @@
     /**
      * Returns a {@link Router} implementation which is using one of {@link RoutingTrie} and {@link List}.
      */
-    private static <V> Router<V> router(boolean isTrie, List<V> values,
+    private static <V> Router<V> router(boolean isTrie, List<V> values, @Nullable V fallbackValue,
                                         Function<V, Route> routeResolver) {
         final Comparator<V> valueComparator =
                 Comparator.comparingInt(e -> -1 * routeResolver.apply(e).complexity());
 
         final Router<V> router;
         if (isTrie) {
-            final RoutingTrie.Builder<V> builder = new Builder<>();
+            final RoutingTrieBuilder<V> builder = new RoutingTrieBuilder<>();
             // Set a comparator to sort services by the number of conditions to be checked in a descending
             // order.
             builder.comparator(valueComparator);
-            values.forEach(v -> builder.add(routeResolver.apply(v).paths().get(1), v));
+            values.forEach(v -> {
+                final Route route = routeResolver.apply(v);
+                final String path = route.paths().get(0);
+                final int pathLen = path.length();
+                if (fallbackValue != null && pathLen > 1) {
+                    // Add some extra routes to handle the case where a client sends a request
+                    // without a trailing slash. `PathMapping.apply()` will produce a `RoutingResult`
+                    // whose type is `NEEDS_REDIRECT`.
+                    switch (route.pathType()) {
+                        case EXACT:         // `/foo`   or `/foo/`
+                        case PARAMETERIZED: // `/foo/:` or `/foo/:/`
+                            if (path.charAt(pathLen - 1) == '/') {
+                                // Add an extra route without a trailing slash for a redirect.
+                                builder.add(path.substring(0, pathLen - 1),
+                                            fallbackValue, /* hasHighPrecedence */ false);
+                            }
+                            break;
+                        case PREFIX: // `/foo/*`
+                            // Add an extra route of an exact match.
+                            builder.add(path.substring(0, pathLen - 1),
+                                        fallbackValue, /* hasHighPrecedence */ false);
+                            break;
+                        default:
+                            throw new Error("Unexpected path type: " + route.pathType());
+                    }
+                }
+                builder.add(route.paths().get(1), v);
```

Comment:
Not sure this is a good idea, but how about just checking the last character if it's `/` or not because the path type is one of `EXACT`, `PARAMETERIZED` and `PREFIX`?
```suggestion
final Route route = routeResolver.apply(v);
builder.add(route.paths().get(1), v);
final String path = route.paths().get(0);
final int pathLen = path.length();
if (fallbackValue != null && pathLen > 1) {
    // Add some extra routes to handle the case where a client sends a request
    // without a trailing slash. `PathMapping.apply()` will produce a `RoutingResult`
    // whose type is `NEEDS_REDIRECT`.
    if (path.charAt(pathLen - 1) == '/') {
        // Add an extra route without a trailing slash for a redirect.
        builder.add(path.substring(0, pathLen - 1),
                    fallbackValue, /* hasHighPrecedence */ false);
    }
}
```

Reply:
I think it simplifies a lot. Thanks for suggestion.